effectful/handlers/llm/providers.py

```diff
@@ -209,11 +289,68 @@ def completion(*args, **kwargs) -> Any:
 
 # Note: attempting to type the tool arguments causes type-checker failures
 @defop
-def tool_call[T](template: Template, tool: Operation[..., T], *args, **kwargs) -> T:
-    """Perform a model-initiated tool call."""
+def tool_call[T](
+    template: Template, tool: Operation[..., T] | Template[..., T], *args, **kwargs
+) -> T:
+    """Perform a model-initiated tool call (can be an Operation or another Template)."""
     return tool(*args, **kwargs)
 
 
+@defop
+def get_tool_depth() -> int:
+    """Get the current tool call depth. Level 0 = initial template call."""
+    return 0
+
+
+@defop
+def get_max_tool_depth() -> int:
+    """Get the maximum allowed tool call depth.
+
+    Default is 0 (no tool calls). Use ToolDepthHandler to enable:
+        handler(ToolDepthHandler(max_depth=1))  # allow 1 level
+    """
+    return 1
+
+
+class ToolDepthHandler(ObjectInterpretation):
```

Comment:
I can understand why `ToolDepthHandler` was introduced, but it's not obvious that it's the correct solution to the motivating problem. My preference would be to remove it for now and just use `RetryHandler` to catch `RecursionError`s, and only introduce new machinery when we have more examples where what's already available is clearly inadequate.

I'd expect stack overflows in `Template` calls to be very rare with good prompts, and perhaps even impossible by construction if we add back annotations that limit the set of available tools in a subsequent PR.

Reply:
To be clear, you can just remove it for now, no need to put it up in a separate PR. I think this particular design is probably not what we want.